Fixes submit button enablement on save/reset

jlajoie opened the pull request:
## [TNL-5619](https://openedx.atlassian.net/browse/TNL-5619)

### Description

Removes the enable/disable all button methods in the display.js file. They did not take into account buttons that were previously disabled, should not be re-enabled after the function was called. The new functions will only modify the buttons passed to them. Tests were updated accordingly.

### Sandbox
- [x] [Sandbox here](https://jlajoie.sandbox.edx.org)

To verify go through the following links:

edX Demonstration Course -> Course -> About Exams and Certificates -> edX Exams -> Limited Checks (has Save) and Randomized Questions (has Reset)

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @staubina 
- [ ] Code review: @cahrens 

FYI: @sstack22 

### Post-review
- [x] Rebase and squash commits

jlajoie commented:
@cahrens @staubina this should be good for review whenever you have a chance

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -36,8 +36,11 @@
             };
-            this.enableAllButtons = function(enable, isFromCheckOperation) { // eslint-disable-line no-unused-vars
-                return Problem.prototype.enableAllButtons.apply(that, arguments);
+            this.enableButtons = function() {
```
This style is a remnant of the conversion from CoffeeScript to JavaScript. Could you change this so that the methods declare their arguments, and then just pass the arguments explicitly to the prototype functions?

jlajoie replied:
Absolutely 👍

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,32 +1160,61 @@
+                    }
+                }
+            });
+            return disabledButtons;
```
I'd like disableButtons and enableButtons to be symmetrical. disableButtons checks that the button is disabled before it does anything, but enableButtons doesn't check that it is enabled. disableButtons also returns something, whereas enableButtons doesn't return the array.

My preference would be to have both methods follow the pattern of enableButtons (assuming there is no harm in disabling something that is already disabled). And then there is no need to return the array of disabled buttons, since the caller already has the array they have called the method with.

cahrens replied:
Oh wait, I just understood the reason for the return value-- that's the whole point of the bug fix. :)

In that case, switch it around so that enableButtons follows the pattern of disableButtons.

jlajoie replied:
Will have them symmetrical, but going to follow your comment below for how the enable and disable button functions should operate.

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,32 +1160,61 @@
-                    attr({disabled: 'disabled'});
-            }
-            return this.enableSubmitButton(enable, isFromCheckOperation);
+        Problem.prototype.disableButtons = function(buttons, changeText) {
```
Since changeText is specific to the submit button, let's name it something that is clearer. How about changeSubmitButtonText?

jlajoie replied:
Easy enough 👍

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,32 +1160,61 @@
             var that = this;
-            this.enableAllButtons(false, isFromCheckOperation);
+            var allButtons = [this.resetButton, this.saveButton, this.showButton, this.hintButton, this.submitButton];
+            var disabledButtons = this.disableButtons(allButtons, isFromCheckOperation);
```
Part of my confusion here is that these aren't all the disabledButtons, they are just the ones that the disableButtons method operated on. I'm trying to think of the clearest way to do this.

It seems there are 2 options-- try to come up with clearer variable names (both here and in disableButtons), or filter "allButtons" to only have the enabled buttons prior to calling disableButtons (in which case disableButtons doesn't need to return anything). I'll let you decide your preference.

jlajoie replied:
I think the easiest route that will make the most sense will be to do the filter operation before calling disableButtons, and then not needing the return.

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -251,15 +251,21 @@ describe 'Problem', ->
           deferred.resolve()
         deferred.promise()
 
+      spyOn(@problem, 'disableButtons').and.callThrough()
```
I'm not a fan of all this spying. Honestly, I don't care which methods were called as long as the state on the individual buttons is correct (which also makes the test more robust to code changes like this one).

I also want to make sure there is a test that fails without the bug fix. I'm not sure that is the case here, since it is just checking which methods are called.

Instead of spying, can you check the disabled attribute on the impacted buttons? This means that you may need to change the test case so that it will have a Save (and possibly Reset) button.

staubina replied:
👍 

I agree with @cahrens It seems like overkill to check for each call to happen.  As long as the later expectation of the button being enabled is met you should be all set.

jlajoie replied:
For this test specifically, it's testing that the buttons are disabled immediately following the submit() call. Submit has both the disabling + enabling of the buttons, so during the execution of submit the button will go from Enabled + Submit, Disabled + Submitting, Enabled + Submit. Is there a way in Jasmine to test that middle state, rather than just start and end? Or is that something that might be better suited as an acceptance test?

jlajoie commented:
@cahrens I think this is ready for another round of feedback. Added tests that would fail with the old code base, removed some of the unnecessary spy calls, updated some variable names, and made the enable/disable consistent with one another. Also removed the coffeescript style function calls

staubina commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,58 +1154,82 @@
-                    attr({disabled: 'disabled'});
-            }
-            return this.enableSubmitButton(enable, isFromCheckOperation);
+        Problem.prototype.disableButtons = function(buttons, changeSubmitButtonText) {
```
These two methods contain mostly the same code can we combine them to reduce duplication?

staubina replied:
Problem.prototype.disableButtons and  Problem.prototype.disableButtons

jlajoie replied:
:+1: good catch, happy to change that

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -251,15 +251,17 @@ describe 'Problem', ->
+      spyOn(@problem, 'enableSubmitButton').and.callThrough()
+      # Submit button should not be disabled after adding an input
+      $('#input_example_1').val('test').trigger('input')
+      expect(@problem.submitButton).not.toHaveAttr('disabled')
```
Can you check before line 260 that the submit button DOES have the disabled attribute?

jlajoie replied:
Yes sounds reasonable

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -251,15 +251,17 @@ describe 'Problem', ->
 
       runs.call(self).then(->
-        expect(self.problem.enableAllButtons).toHaveBeenCalledWith true, true
+        expect(self.problem.enableSubmitButton).toHaveBeenCalledWith false, true
```
Why is it necessary to use this spy statement instead of checking the disabled state of the submit button?

jlajoie replied:
Going to go through and investigate the point of deferred, as well as see if it is possible to check the actual state of the button rather than the parameters of the function with a spy.

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -251,15 +251,17 @@ describe 'Problem', ->
         if jQuery.active == 0
           deferred.resolve()
         deferred.promise()
+      spyOn(@problem, 'enableSubmitButton').and.callThrough()
```
The description of this test case (on line 239) is no longer accurate.

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -275,8 +277,11 @@ describe 'Problem', ->
               callable()
-        spyOn @problem.submitButtonLabel, 'text'
+        spyOn(@problem.submitButtonLabel, 'text')
+        $('#input_example_1').val('test').trigger('input')
```
It seems like this test is covering the same sort of things (with regard to enable/disable state) as the previous test point. Do we need 2 different test points? 

If it makes sense to keep a separate test around text changing (which I can certainly be on board with), I'd recommend removing the enable/disable checks from this test point, as the coverage should be in the test point that is specifically about enable/disable state.

jlajoie replied:
Sounds good, happy to keep the logic separate, will remove the unneeded calls for disabled/enabled.

staubina commented:
👍

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -424,6 +429,47 @@ describe 'Problem', ->
+          deferred.resolve()
+        deferred.promise()
+
+      runs.call(self).then(->
```
I'm not convinced this test point is working correctly. Since the spyOn is defined in the "runs", and you call reset before you ever call "runs", how does the spy actually do anything? Shouldn't the spyOn be handing the ajax call for runs?

I think this could all be simplified somewhat. Take a look at the specific "reset" test points above.

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -275,8 +277,11 @@ describe 'Problem', ->
+        $('#input_example_1').val('test').trigger('input')
+        expect(@problem.submitButton).not.toHaveAttr('disabled')
         @problem.submit()
+        expect(@problem.submitButton).not.toHaveAttr('disabled')
```
Should be disabled.

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,58 +1154,64 @@
 
         /**
-         * Used to enable/disable all buttons in problem.
+         * Enables all buttons by removing the disabled attribute. The submit button is checked separately due to
```
This documentation is no longer correct since you went back to a single disable/enable method.

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,58 +1154,64 @@
         Problem.prototype.disableAllButtonsWhileRunning = function(operationCallback, isFromCheckOperation) {
             var that = this;
-            this.enableAllButtons(false, isFromCheckOperation);
+            var allButtons = [this.resetButton, this.saveButton, this.showButton, this.hintButton, this.submitButton];
```
I like this much better.

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,58 +1154,64 @@
          *         text of submit button as well.
          */
-        Problem.prototype.enableSubmitButton = function(enable, changeText) {
+        Problem.prototype.enableSubmitButton = function(enable, changeSubmitButtonText) {
```
Nit: you could keep this just "changeText" in this method because the method itself is scoped to the Submit button.

jlajoie replied:
Of course :+1:

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,58 +1154,64 @@
+            var initiallyEnabledButtons = allButtons.filter(function(button) {
+                return !button.attr('disabled');
+            });
+            this.enableButtons(initiallyEnabledButtons, false, isFromCheckOperation);
```
So enableButtons is no longer defined on the class instance itself, just the prototype (since you deleted lines 39-41 above). But calling it here works because this is already a prototype method?

jlajoie replied:
Yes, if we wanted to keep the Coffeescript style, we would still need to follow the pattern of using apply for the functions rather than just passing the variables in directly, or else we would lose the execution context. Is there a preference to not having them declared in the `function Problem()` vs having them declared and using apply?

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -723,6 +769,48 @@ describe 'Problem', ->
+      deferred = $.Deferred()
+      self = this
+      curr_html = @problem.el.html()
+      runs = ->
```
There is so much boilerplate in these test points.... If they can't be reduced down to something simple, please at least declare a common "postWithPrefix" spy that can be reused between test points.

jlajoie replied:
Happy to do so, will investigate if a lot of this boilerplate is even needed

jlajoie commented:
This should be good for another pass whenever you have a chance @staubina @cahrens

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.js:
```diff
@@ -1157,32 +1154,39 @@
+            var that = this;
+            buttons.forEach(function(button) {
+                if (button.hasClass('submit')) {
+                    console.log('calling enable submit with ' + enable);
```
Don't forget to remove console.log.

staubina commented:
👍

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -236,56 +236,27 @@ describe 'Problem', ->
+        # At this point enableButtons should have been called, making the submit button disabled with text 'submitting'
+        expect(self.problem.submitButton).toHaveAttr('disabled');
+        expect(self.problem.submitButtonLabel.text()).toBe('Submitting');
+        promise = undefined
```
Nit: why do you set promise to undefined here?

jlajoie replied:
I was getting a linting error from my IDE, `unused local variable promise` if I didn't initialize it. Though I did just notice there's other tests I didn't touch that have this linting error that pass our own style checks. I'll go through and update these to make them consistent, then this should be good to go?

cahrens replied:
Yes, I don't need to look at it again.

cahrens commented:
👍 I had one question about why a variable was initialized to undefined, but it's to you if you wish to change it.
